### ISSUE TITLE
Consolidate package exceptions

### ISSUE THREAD

**Issue by rocco8773** [PlasmaPy/PlasmaPy#881]
I feel it's probably a good idea to have all PlasmaPy defined Exceptions (and Warnings) be accessible from one place (e.g. `plasmapy.utils.exceptions`).  Using `plasmapy.utils.exceptions` as an example, this could be done one of two ways: (1) have all exception definitions in `plasmapy.utils.exceptions` or (2) leave the definitions where they are and import them into the `plasmapy.utils.exceptions` namespace.  Thoughts?  Option 2 is probably the better way of doing it, well, at least easier to implement.

**Comment by StanczakDominik** [comment]
Eh... I don't think it is that hard to grep for the exceptions, and I suspect pycharm etc would find them wherever they are if you just find PlasmaPyError first.

Besides, on a practical note -say I define relativity error in relativity.py, as that's where that exception is used. Then, with this idea, I'd import it for utils. But I will then run into a circular dependency as I had already imported PlasmaPyError to subclass.

**Comment by rocco8773** [comment]
I agree with you.  However, I don't think it's that straightforward for "non-expert" users.  I think having one home for exceptions would be more intuitive for those users in the early stages of adoption.

Besides `plasmapy.roman.roman`, I think all of our exceptions are defined in an `exceptions.py` in the relevant directory.  `RelativityError` is already defined in `plasmapy.utils.exceptions`.  I do think this has been a good practice thus far.

You're right we'd have to avoid circular imports.  I think it could be done though.

**Comment by StanczakDominik** [comment]
Okay then, I forgot we moved them all there already. Sorry!

**Comment by namurphy** [comment]
Here are my current thoughts with regards to how we organize exceptions.

1. We put the broadest exceptions for PlasmaPy into `plasmapy.utils.exceptions`.  
  a. Examples: `PlasmaPyError`, `PlasmaPyWarning`, `PhysicsError`, etc.
  b. This module would include base exceptions and exceptions used across multiple packages.
2. Each subpackage should have `exceptions.py`, which contains the exceptions for that subpackage.
  a. This could possibly also be done in subpackages of subpackages...though at the moment I only think this would be helpful in `plasmapy.utils`.
3. We only import `exceptions` into the top-level of each subpackage, not the specific exceptions.
  a. Example: we'd import `plasmapy.utils.exceptions.PlasmaPyError` rather than `plasmapy.utils.PlasmaPyError`.  
  b. This would reduce namespace pollution.
4. We could possibly also have `warnings.py`
  a.  Probably not necessary yet since warnings inherit from `BaseException`.  

I'm still open to other possibilities, so long as we treat exceptions in a consistent and obvious way.  It'd also be helpful to update the development guide to reflect what we decide on.
